// @appium/docutils/build/lib/utils/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.stopwatch = exports.spawnBackgroundProcess = exports.execWithErrorHandling = exports.relative = exports.mergeDefaultsDeep = exports.isStringArray = exports.kebabCase = exports.argify = exports.readPackage = exports.findPackageRootSync = exports.findPackageRoot = void 0;
var package_json_1 = require("./package-json");
Object.defineProperty(exports, "findPackageRoot", { enumerable: true, get: function () { return package_json_1.findPackageRoot; } });
Object.defineProperty(exports, "findPackageRootSync", { enumerable: true, get: function () { return package_json_1.findPackageRootSync; } });
Object.defineProperty(exports, "readPackage", { enumerable: true, get: function () { return package_json_1.readPackage; } });
var cli_1 = require("./cli");
Object.defineProperty(exports, "argify", { enumerable: true, get: function () { return cli_1.argify; } });
Object.defineProperty(exports, "kebabCase", { enumerable: true, get: function () { return cli_1.kebabCase; } });
var object_1 = require("./object");
Object.defineProperty(exports, "isStringArray", { enumerable: true, get: function () { return object_1.isStringArray; } });
Object.defineProperty(exports, "mergeDefaultsDeep", { enumerable: true, get: function () { return object_1.mergeDefaultsDeep; } });
var path_1 = require("./path");
Object.defineProperty(exports, "relative", { enumerable: true, get: function () { return path_1.relative; } });
var process_1 = require("./process");
Object.defineProperty(exports, "execWithErrorHandling", { enumerable: true, get: function () { return process_1.execWithErrorHandling; } });
Object.defineProperty(exports, "spawnBackgroundProcess", { enumerable: true, get: function () { return process_1.spawnBackgroundProcess; } });
var timing_1 = require("./timing");
Object.defineProperty(exports, "stopwatch", { enumerable: true, get: function () { return timing_1.stopwatch; } });
