// @appium/base-driver/build/lib/express/express-logging.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.startLogFormatter = exports.endLogFormatter = void 0;
const support_1 = require("@appium/support");
const morgan_1 = __importDefault(require("morgan"));
const logger_1 = require("./logger");
const constants_1 = require("../constants");
/**
 * Morgan middleware that logs when the HTTP response finishes.
 * Logs method, URL, status (color-coded), response time, and content-length.
 */
exports.endLogFormatter = (0, morgan_1.default)(endLogFormatterHandler);
/**
 * Morgan middleware that logs when the HTTP request is received (immediate).
 * Logs method and URL; request body is truncated and passed through {@link logger.markSensitive}.
 */
exports.startLogFormatter = (0, morgan_1.default)(startLogFormatterHandler, {
    immediate: true,
});
function endLogFormatterHandler(tokens, req, res) {
    logger_1.log.info(requestEndLoggingFormat(tokens, req, res));
    return undefined;
}
function startLogFormatterHandler(tokens, req, res) {
    let reqBody = '';
    if (req.body) {
        try {
            reqBody = support_1.util.truncateString(typeof req.body === 'string' ? req.body : JSON.stringify(req.body), { length: constants_1.MAX_LOG_BODY_LENGTH });
        }
        catch {
            // ignore
        }
    }
    logger_1.log.info(requestStartLoggingFormat(tokens, req, res), support_1.logger.markSensitive(support_1.console.styleText('grey', reqBody)));
    return undefined;
}
// Copied the morgan compile function over so that cooler formats may be configured
function compile(fmt) {
    fmt = fmt.replace(/"/g, '\\"');
    fmt = fmt.replace(/:([-\w]{2,})(?:\[([^\]]+)\])?/g, function replace(_, name, arg) {
        return `"\n    + (tokens["${name}"](req, res, "${arg}") || "-") + "`;
    });
    const js = `  return "${fmt}";`;
    return new Function('tokens', 'req', 'res', js);
}
function requestEndLoggingFormat(tokens, req, res) {
    const status = res.statusCode;
    let statusStr = ':status';
    if (status >= 500) {
        statusStr = support_1.console.styleText('red', statusStr);
    }
    else if (status >= 400) {
        statusStr = support_1.console.styleText('yellow', statusStr);
    }
    else if (status >= 300) {
        statusStr = support_1.console.styleText('cyan', statusStr);
    }
    else {
        statusStr = support_1.console.styleText('green', statusStr);
    }
    const fn = compile(`${support_1.console.styleText('white', '<-- :method :url ')}${statusStr} ${support_1.console.styleText('grey', ':response-time ms - :res[content-length]')}`);
    return fn(tokens, req, res);
}
const requestStartLoggingFormat = compile(`${support_1.console.styleText('white', '-->')} ${support_1.console.styleText('white', ':method')} ${support_1.console.styleText('white', ':url')}`);
// #endregion
