// appium/build/lib/doctor/doctor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Doctor = exports.EXIT_CODE = void 0;
const support_1 = require("@appium/support");
/**
 * Process exit codes returned by {@link Doctor.run}.
 */
exports.EXIT_CODE = Object.freeze({
    SUCCESS: 0,
    HAS_MAJOR_ISSUES: 127,
});
class Doctor {
    log;
    checks;
    foundIssues;
    constructor(checks = []) {
        this.log = support_1.logger.getLogger('Doctor');
        this.checks = checks;
        this.checks
            .filter((c) => !support_1.util.hasValue(c.log))
            .forEach((c) => {
            c.log = this.log;
        });
        this.foundIssues = [];
    }
    get issuesRequiredToFix() {
        return this.foundIssues.filter((f) => !f.check.isOptional());
    }
    get issuesOptionalToFix() {
        return this.foundIssues.filter((f) => f.check.isOptional());
    }
    /**
     * Runs diagnostics, reports issues, attempts automatic fixes where supported, and returns an exit code.
     *
     * @returns {@link EXIT_CODE.SUCCESS} when there are no issues or all issues were resolved;
     *   {@link EXIT_CODE.HAS_MAJOR_ISSUES} when manual intervention is still required or fixes failed.
     */
    async run() {
        await this.diagnose();
        if (this.reportSuccess()) {
            return exports.EXIT_CODE.SUCCESS;
        }
        if (await this.reportManualIssues()) {
            return exports.EXIT_CODE.HAS_MAJOR_ISSUES;
        }
        if (!(await this.runAutoFixes())) {
            return exports.EXIT_CODE.HAS_MAJOR_ISSUES;
        }
        return exports.EXIT_CODE.SUCCESS;
    }
    /**
     * The doctor shows the report
     */
    async diagnose() {
        this.log.info(`### Starting doctor diagnostics  ###`);
        this.foundIssues = [];
        for (const check of this.checks) {
            const res = await check.diagnose();
            const issue = this.toIssue(res, check);
            if (issue) {
                this.foundIssues.push(issue);
            }
        }
        this.log.info(`### Diagnostic completed, ${this.buildFixMessage()}. ###`);
        this.log.info('');
    }
    async reportManualIssues() {
        const manualIssues = this.issuesRequiredToFix.filter((f) => !f.check.hasAutofix());
        const manualIssuesOptional = this.issuesOptionalToFix.filter((f) => !f.check.hasAutofix());
        const handleIssues = async (headerLogs, issues) => {
            if (support_1.util.isEmpty(issues)) {
                return;
            }
            for (const logMsg of headerLogs) {
                this.log.info(logMsg);
            }
            const fixMessages = [];
            for (const issue of issues) {
                let message;
                try {
                    message = await issue.check.fix();
                }
                catch (e) {
                    message = e.message;
                }
                if (message) {
                    fixMessages.push(message);
                }
            }
            for (const m of support_1.util.uniq(fixMessages)) {
                this.log.warn(` \u279C ${m}`);
            }
            this.log.info('');
        };
        await handleIssues([
            '### Manual Fixes Needed ###',
            'The configuration cannot be automatically fixed, please do the following first:',
        ], manualIssues);
        await handleIssues([
            '### Optional Manual Fixes ###',
            'To fix these optional issues, please do the following manually:',
        ], manualIssuesOptional);
        if (manualIssues.length > 0) {
            this.log.info('###');
            this.log.info('');
            this.log.info('Bye! Run doctor again when all manual fixes have been applied!');
            this.log.info('');
            return true;
        }
        return false;
    }
    async runAutoFix(f) {
        this.log.info(`### Fixing: ${f.error} ###`);
        try {
            await f.check.fix();
        }
        catch (err) {
            if (err instanceof support_1.doctor.FixSkippedError) {
                this.log.info(`### Skipped fix ###`);
                return;
            }
            else {
                this.log.warn(`${err}`.replace(/\n$/g, ''));
                this.log.info(`### Fix did not succeed ###`);
                return;
            }
        }
        this.log.info('Checking if this was fixed:');
        const res = await f.check.diagnose();
        if (res.ok) {
            f.fixed = true;
            this.log.info(` ${support_1.console.styleText('green', '\u2714')} ${res.message}`);
            this.log.info(`### Fix was successfully applied ###`);
        }
        else {
            this.log.info(` ${support_1.console.styleText('red', '\u2716')} ${res.message}`);
            this.log.info(`### Fix was applied but issue remains ###`);
        }
    }
    async runAutoFixes() {
        const autoFixes = this.foundIssues.filter((f) => f.check.hasAutofix());
        for (const f of autoFixes) {
            await this.runAutoFix(f);
            this.log.info('');
        }
        if (autoFixes.find((f) => !f.fixed)) {
            // a few issues remain.
            this.log.info('Bye! A few issues remain, fix manually and/or rerun doctor!');
            this.log.info('');
            return false;
        }
        // nothing left to fix.
        this.log.info('Bye! All issues have been fixed!');
        this.log.info('');
        return true;
    }
    toIssue(result, check) {
        if (result.ok) {
            this.log.info(` ${support_1.console.styleText('green', '\u2714')} ${result.message}`);
            return null;
        }
        const errorMessage = result.optional
            ? ` ${support_1.console.styleText('yellow', '\u2716')} ${result.message}`
            : ` ${support_1.console.styleText('red', '\u2716')} ${result.message}`;
        this.log.warn(errorMessage);
        return {
            error: errorMessage,
            check,
        };
    }
    buildFixMessage() {
        return (`${support_1.util.pluralize('required fix', this.issuesRequiredToFix.length, true)} needed, ` +
            `${support_1.util.pluralize('optional fix', this.issuesOptionalToFix.length, true)} possible`);
    }
    reportSuccess() {
        if (this.issuesRequiredToFix.length === 0 && this.issuesOptionalToFix.length === 0) {
            this.log.info('Everything looks good, bye!');
            this.log.info('');
            return true;
        }
        return false;
    }
}
exports.Doctor = Doctor;
