// appium/build/lib/utils/is-package-changed.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.isPackageChanged = isPackageChanged;
const support_1 = require("@appium/support");
const node_crypto_1 = __importDefault(require("node:crypto"));
const node_path_1 = __importDefault(require("node:path"));
/**
 * Detects whether `package.json` dependencies changed since the last hash write.
 * Inlined from the `package-changed` package (lockfile support omitted; unused by Appium).
 */
async function isPackageChanged(options = {}) {
    const { hashFilename = '.packagehash', cwd = process.cwd() } = options;
    const packagePath = await findPackageJson(cwd);
    if (!packagePath) {
        throw new Error(`Cannot find package.json travelling up from "${cwd}".`);
    }
    const packageHashPath = node_path_1.default.join(cwd, hashFilename);
    const packageHashPathExists = await support_1.fs.exists(packageHashPath);
    const recentDigest = await getPackageHash(packagePath);
    const previousDigest = packageHashPathExists
        ? await support_1.fs.readFile(packageHashPath, 'utf-8')
        : undefined;
    const isChanged = !packageHashPathExists || previousDigest !== recentDigest;
    const writeHash = async () => {
        await support_1.fs.writeFile(packageHashPath, recentDigest, 'utf-8');
    };
    return {
        hash: recentDigest,
        isChanged,
        oldHash: previousDigest || undefined,
        writeHash,
    };
}
async function findPackageJson(cwd) {
    let current = cwd;
    while (true) {
        const search = node_path_1.default.join(current, 'package.json');
        if (await support_1.fs.exists(search)) {
            return search;
        }
        const parent = node_path_1.default.dirname(current);
        if (parent === current) {
            return;
        }
        current = parent;
    }
}
async function getPackageHash(packagePath) {
    const contents = await support_1.fs.readFile(packagePath, 'utf-8');
    const packageBlob = JSON.parse(contents);
    const dependencies = {
        dependencies: packageBlob.dependencies ?? {},
        devDependencies: packageBlob.devDependencies ?? {},
    };
    const hashSum = node_crypto_1.default.createHash('md5');
    hashSum.update(Buffer.from(JSON.stringify(dependencies)));
    return hashSum.digest('hex');
}
