// appium/build/lib/utils/object.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.kebabCase = kebabCase;
exports.camelCase = camelCase;
exports.capitalize = capitalize;
exports.omitKeys = omitKeys;
exports.pickBy = pickBy;
exports.mapValues = mapValues;
exports.mapKeys = mapKeys;
exports.getPath = getPath;
exports.setPath = setPath;
exports.bindAll = bindAll;
exports.compact = compact;
exports.pull = pull;
exports.zip = zip;
exports.difference = difference;
exports.defaultsDeep = defaultsDeep;
const support_1 = require("@appium/support");
const UNSAFE_PATH_KEYS = new Set(['__proto__', 'constructor', 'prototype']);
/**
 * Converts a string to kebab-case (for Appium CLI and schema property names).
 *
 * @param value - Input string
 * @returns Kebab-cased string
 */
function kebabCase(value) {
    return splitWords(value).join('-').toLowerCase();
}
/**
 * Converts a string to camelCase.
 *
 * @param value - Input string
 * @returns camelCased string
 */
function camelCase(value) {
    const words = splitWords(value);
    if (words.length === 0) {
        return '';
    }
    return (words[0].toLowerCase() +
        words
            .slice(1)
            .map((w) => w.charAt(0).toUpperCase() + w.slice(1).toLowerCase())
            .join(''));
}
/**
 * Uppercases the first character of a string.
 *
 * @param value - Input string
 * @returns Capitalized string, or empty string when input is empty
 */
function capitalize(value) {
    return value ? value.charAt(0).toUpperCase() + value.slice(1) : value;
}
/**
 * Returns a shallow copy of `obj` without any of `keys`. Non-plain objects are returned unchanged.
 *
 * @param obj - Source object
 * @param keys - Property names to omit
 * @returns Shallow copy without the listed keys
 */
function omitKeys(obj, keys) {
    if (!support_1.util.isPlainObject(obj) || keys.length === 0) {
        return obj;
    }
    const omit = new Set(keys);
    return Object.fromEntries(Object.entries(obj).filter(([k]) => !omit.has(k)));
}
/**
 * Returns a shallow copy of `obj` whose entries pass `predicate`.
 *
 * @param obj - Source object
 * @param predicate - Filter invoked with each value and key
 * @returns Shallow copy containing only matching entries
 */
function pickBy(obj, predicate) {
    return Object.fromEntries(Object.entries(obj).filter(([key, value]) => predicate(value, key)));
}
/**
 * Returns an object with the same keys as `obj` and values transformed by `fn`.
 *
 * @param obj - Source object
 * @param fn - Mapper invoked with each value and key
 * @returns Object with transformed values
 */
function mapValues(obj, fn) {
    return Object.fromEntries(Object.entries(obj).map(([key, value]) => [key, fn(value, key)]));
}
/**
 * Returns an object with keys renamed by `fn` and original values preserved.
 *
 * @param obj - Source object
 * @param fn - Key mapper invoked with each value and key
 * @returns Object with renamed keys
 */
function mapKeys(obj, fn) {
    return Object.fromEntries(Object.entries(obj).map(([key, value]) => [fn(value, key), value]));
}
/**
 * Reads a dot-separated path on `obj`, returning `defaultValue` when any segment is missing.
 *
 * @param obj - Object to read from
 * @param path - Dot-separated property path
 * @param defaultValue - Value returned when the path cannot be resolved
 * @returns Value at the path, or `defaultValue`
 */
function getPath(obj, path, defaultValue) {
    const parts = path.split('.');
    let current = obj;
    for (const part of parts) {
        if (current == null || (typeof current !== 'object' && typeof current !== 'function')) {
            return defaultValue;
        }
        current = current[part];
    }
    return current === undefined ? defaultValue : current;
}
/**
 * Assigns `value` at a dot-separated path on `obj`, creating plain object segments as needed.
 *
 * No-ops when the path contains empty segments or keys that could trigger prototype pollution
 * (`__proto__`, `constructor`, `prototype`).
 *
 * @param obj - Object to mutate
 * @param path - Dot-separated property path
 * @param value - Value to assign at the path
 */
function setPath(obj, path, value) {
    const parts = path.split('.');
    if (!parts.every(isSafePathKey)) {
        return;
    }
    let current = obj;
    for (let i = 0; i < parts.length - 1; i++) {
        const part = parts[i];
        const next = current[part];
        if (!support_1.util.isPlainObject(next)) {
            current[part] = {};
        }
        current = current[part];
    }
    current[parts[parts.length - 1]] = value;
}
/**
 * Binds listed methods on `obj` so they keep the correct `this` when passed as callbacks.
 *
 * @param obj - Target object
 * @param methodNames - Method names to bind
 * @returns `obj` (mutated in place)
 */
function bindAll(obj, methodNames) {
    for (const name of methodNames) {
        const method = obj[name];
        if (typeof method === 'function') {
            obj[name] = method.bind(obj);
        }
    }
    return obj;
}
/**
 * Returns a copy of `arr` with falsy entries removed.
 *
 * @param arr - Input array
 * @returns Array containing only truthy elements
 */
function compact(arr) {
    return arr.filter((x) => Boolean(x));
}
/**
 * Removes all occurrences of `values` from `arr` (mutates `arr`).
 *
 * @param arr - Array to modify
 * @param values - Values to remove
 * @returns `arr`
 */
function pull(arr, ...values) {
    for (const v of values) {
        let idx;
        while ((idx = arr.indexOf(v)) !== -1) {
            arr.splice(idx, 1);
        }
    }
    return arr;
}
/**
 * Zips two arrays into tuples `[a[i], b[i]]`.
 *
 * @param a - First array
 * @param b - Second array
 * @returns Array of zipped pairs
 */
function zip(a, b) {
    return a.map((item, i) => [item, b[i]]);
}
/**
 * Returns elements in `a` that are not present in `b`.
 *
 * @param a - Source array
 * @param b - Values to exclude
 * @returns Filtered array
 */
function difference(a, b) {
    const bSet = new Set(b);
    return a.filter((x) => !bSet.has(x));
}
/**
 * Deep-defaults merge: each source fills only `undefined` properties in the result (recursive for plain objects).
 *
 * @param sources - Objects merged in order; null/undefined sources are skipped
 * @returns Merged object
 */
function defaultsDeep(...sources) {
    let result = {};
    for (const source of sources) {
        if (source == null) {
            continue;
        }
        result = fillUndefinedDeep(result, source);
    }
    return result;
}
/**
 * Splits a string into words (lodash-compatible), including acronym boundaries.
 *
 * @param value - Input string
 * @returns Word tokens
 */
function splitWords(value) {
    return value
        .replace(/([A-Z]+)([A-Z][a-z])/g, '$1 $2')
        .replace(/([a-z0-9])([A-Z])/g, '$1 $2')
        .replace(/[-_\s]+/g, ' ')
        .trim()
        .split(/\s+/)
        .filter(Boolean);
}
function isSafePathKey(key) {
    return key.length > 0 && !UNSAFE_PATH_KEYS.has(key);
}
/**
 * Copies a value for {@link defaultsDeep}: plain objects are copied recursively;
 * other values (including functions) are reused by reference.
 *
 * @param value - Value to copy
 * @returns Copy safe for further defaults merging
 */
function copyForDefaultsDeep(value) {
    if (!support_1.util.isPlainObject(value)) {
        return value;
    }
    const out = {};
    for (const [key, val] of Object.entries(value)) {
        out[key] = copyForDefaultsDeep(val);
    }
    return out;
}
/**
 * Fills `undefined` keys in a clone of `target` from `source` (recursive for plain objects).
 *
 * @param target - Destination object
 * @param source - Default values
 * @returns Merged clone
 */
function fillUndefinedDeep(target, source) {
    const out = copyForDefaultsDeep(target);
    const stack = [
        { dest: out, src: source },
    ];
    while (stack.length) {
        const next = stack.pop();
        if (!next) {
            continue;
        }
        const { dest, src } = next;
        for (const [key, srcVal] of Object.entries(src)) {
            const destVal = dest[key];
            if (destVal === undefined) {
                dest[key] = copyForDefaultsDeep(srcVal);
                continue;
            }
            if (support_1.util.isPlainObject(destVal) && support_1.util.isPlainObject(srcVal)) {
                stack.push({ dest: destVal, src: srcVal });
            }
        }
    }
    return out;
}
