// @appium/logger/build/lib/utils/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setBlocking = exports.unleakString = exports.isPlainObject = exports.escapeRegExp = exports.ansiColor = exports.ansiBeep = void 0;
var ansi_1 = require("./ansi");
Object.defineProperty(exports, "ansiBeep", { enumerable: true, get: function () { return ansi_1.ansiBeep; } });
Object.defineProperty(exports, "ansiColor", { enumerable: true, get: function () { return ansi_1.ansiColor; } });
var node_1 = require("./node");
Object.defineProperty(exports, "escapeRegExp", { enumerable: true, get: function () { return node_1.escapeRegExp; } });
Object.defineProperty(exports, "isPlainObject", { enumerable: true, get: function () { return node_1.isPlainObject; } });
Object.defineProperty(exports, "unleakString", { enumerable: true, get: function () { return node_1.unleakString; } });
var set_blocking_1 = require("./set-blocking");
Object.defineProperty(exports, "setBlocking", { enumerable: true, get: function () { return set_blocking_1.setBlocking; } });
