// appium/build/lib/inspector-commands.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.listCommands = listCommands;
exports.listExtensions = listExtensions;
const utils_1 = require("./utils");
const base_driver_1 = require("@appium/base-driver");
/**
 * Returns available REST and BiDi commands for base, driver and plugins.
 */
async function listCommands(sessionId) {
    let driverRestMethodMap = {};
    let driverBiDiCommands = {};
    let pluginRestMethodMaps = {};
    let pluginBiDiCommands = {};
    if (sessionId) {
        const driverClass = this.driverForSession(sessionId)?.constructor;
        driverRestMethodMap = driverClass?.newMethodMap ?? {};
        driverBiDiCommands = driverClass?.newBidiCommands ?? {};
        const pluginClasses = this.pluginsForSession(sessionId).map((p) => p.constructor);
        pluginRestMethodMaps = Object.fromEntries(pluginClasses.map((c) => [c.name, c.newMethodMap ?? {}]));
        pluginBiDiCommands = Object.fromEntries(pluginClasses.map((c) => [c.name, c.newBidiCommands ?? {}]));
    }
    return {
        rest: {
            base: methodMapToRestCommandsInfo(base_driver_1.METHOD_MAP),
            driver: methodMapToRestCommandsInfo(driverRestMethodMap),
            plugins: pluginRestMethodMaps
                ? (0, utils_1.mapValues)(pluginRestMethodMaps, methodMapToRestCommandsInfo)
                : undefined,
        },
        bidi: toBiDiCommandsMap(base_driver_1.BIDI_COMMANDS, driverBiDiCommands, pluginBiDiCommands),
    };
}
/**
 * Returns available execute methods exposed by driver and plugins.
 */
async function listExtensions(sessionId) {
    let driverExecuteMethodMap = {};
    let pluginExecuteMethodMaps = {};
    if (sessionId) {
        const driverClass = this.driverForSession(sessionId)?.constructor;
        driverExecuteMethodMap = driverClass?.executeMethodMap ?? {};
        const pluginClasses = this.pluginsForSession(sessionId).map((p) => p.constructor);
        pluginExecuteMethodMaps = Object.fromEntries(pluginClasses.map((c) => [c.name, c.executeMethodMap ?? {}]));
    }
    return {
        rest: {
            driver: executeMethodMapToCommandsInfo(driverExecuteMethodMap),
            plugins: pluginExecuteMethodMaps
                ? (0, utils_1.mapValues)(pluginExecuteMethodMaps, executeMethodMapToCommandsInfo)
                : undefined,
        },
    };
}
function toRestCommandParams(params) {
    if (!params) {
        return;
    }
    const toRestCommandItemParam = (x, isRequired) => {
        const isNameAnArray = Array.isArray(x);
        const name = isNameAnArray ? x[0] : x;
        if (typeof name !== 'string') {
            return;
        }
        // If parameter names are arrays then this means
        // either of them is required.
        // Not sure we could reflect that in here.
        const required = isRequired && !isNameAnArray;
        return {
            name,
            required,
        };
    };
    const requiredParams = (params.required ?? [])
        .map((name) => toRestCommandItemParam(name, true))
        .filter((x) => x !== undefined);
    const optionalParams = (params.optional ?? [])
        .map((name) => toRestCommandItemParam(name, false))
        .filter((x) => x !== undefined);
    return requiredParams.length || optionalParams.length
        ? [...requiredParams, ...optionalParams]
        : undefined;
}
function methodMapToRestCommandsInfo(mm) {
    const res = {};
    for (const [uriPath, methods] of Object.entries(mm)) {
        const methodsMap = {};
        for (const [method, spec] of Object.entries(methods)) {
            methodsMap[String(method)] = {
                command: spec.command,
                deprecated: spec.deprecated,
                info: spec.info,
                params: toRestCommandParams(spec.payloadParams),
            };
        }
        res[uriPath] = methodsMap;
    }
    return res;
}
function executeMethodMapToCommandsInfo(emm) {
    const result = {};
    for (const [name, info] of Object.entries(emm)) {
        result[name] = {
            command: info.command,
            deprecated: info.deprecated,
            info: info.info,
            params: toRestCommandParams(info.params),
        };
    }
    return result;
}
function toBiDiCommandsMap(baseModuleMap, driverModuleMap, pluginModuleMaps) {
    const toBiDiCommandParams = (params) => {
        if (!params) {
            return;
        }
        const toBiDiCommandItemParam = (x, isRequired) => {
            const isNameAnArray = Array.isArray(x);
            const name = isNameAnArray ? x[0] : x;
            if (typeof name !== 'string') {
                return;
            }
            // If parameter names are arrays then this means
            // either of them is required.
            // Not sure we could reflect that in here.
            const required = isRequired && !isNameAnArray;
            return {
                name,
                required,
            };
        };
        const requiredParams = (params.required ?? [])
            .map((name) => toBiDiCommandItemParam(name, true))
            .filter((x) => x !== undefined);
        const optionalParams = (params.optional ?? [])
            .map((name) => toBiDiCommandItemParam(name, false))
            .filter((x) => x !== undefined);
        return requiredParams.length || optionalParams.length
            ? [...requiredParams, ...optionalParams]
            : undefined;
    };
    const moduleMapToBiDiCommandsInfo = (mm) => {
        const res = {};
        for (const [domain, commands] of Object.entries(mm)) {
            const commandsMap = {};
            for (const [name, spec] of Object.entries(commands)) {
                commandsMap[name] = {
                    command: spec.command,
                    deprecated: spec.deprecated,
                    info: spec.info,
                    params: toBiDiCommandParams(spec.params),
                };
            }
            res[domain] = commandsMap;
        }
        return res;
    };
    return {
        base: moduleMapToBiDiCommandsInfo(baseModuleMap),
        driver: moduleMapToBiDiCommandsInfo(driverModuleMap),
        plugins: pluginModuleMaps
            ? (0, utils_1.mapValues)(pluginModuleMaps, moduleMapToBiDiCommandsInfo)
            : undefined,
    };
}
