// appium/build/lib/helpers/capability.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeNonW3cCapsError = makeNonW3cCapsError;
exports.parseCapsForInnerDriver = parseCapsForInnerDriver;
exports.insertAppiumPrefixes = insertAppiumPrefixes;
exports.removeAppiumPrefixes = removeAppiumPrefixes;
exports.pullSettings = pullSettings;
const support_1 = require("@appium/support");
const logger_1 = require("../logger");
const base_driver_1 = require("@appium/base-driver");
const utils_1 = require("../utils");
const W3C_APPIUM_PREFIX = 'appium';
const STANDARD_CAPS_LOWERCASE = new Set([...base_driver_1.STANDARD_CAPS].map((cap) => cap.toLowerCase()));
/**
 * Creates an error when a session receives non-W3C capabilities.
 */
function makeNonW3cCapsError() {
    return new base_driver_1.errors.SessionNotCreatedError('Session capabilities format must comply to the W3C standard. Make sure your client is up to date. ' +
        'See https://www.w3.org/TR/webdriver/#new-session for more details.');
}
/**
 * Parses W3C capabilities for the inner driver and applies defaults.
 *
 * @returns Parsed caps or an invalid result with an error.
 */
function parseCapsForInnerDriver(w3cCapabilities, constraints = {}, defaultCapabilities = {}) {
    if (!(0, base_driver_1.isW3cCaps)(w3cCapabilities)) {
        return { error: makeNonW3cCapsError() };
    }
    let desiredCaps = {};
    // eslint-disable-next-line prefer-const -- assigned in success path after try
    let processedW3CCapabilities;
    w3cCapabilities = structuredClone(w3cCapabilities);
    defaultCapabilities = structuredClone(defaultCapabilities);
    if (!support_1.util.isEmpty(defaultCapabilities)) {
        for (const [defaultCapKey, defaultCapValue] of Object.entries(defaultCapabilities)) {
            let isCapAlreadySet = false;
            for (const firstMatchEntry of w3cCapabilities.firstMatch ?? []) {
                if (support_1.util.isPlainObject(firstMatchEntry) &&
                    hasOwnCapability(firstMatchEntry, defaultCapKey)) {
                    isCapAlreadySet = true;
                    break;
                }
            }
            isCapAlreadySet =
                isCapAlreadySet ||
                    (support_1.util.isPlainObject(w3cCapabilities.alwaysMatch) &&
                        hasOwnCapability(w3cCapabilities.alwaysMatch, defaultCapKey));
            if (isCapAlreadySet) {
                continue;
            }
            if (support_1.util.isEmpty(w3cCapabilities.firstMatch)) {
                w3cCapabilities.firstMatch = [
                    { [defaultCapKey]: defaultCapValue },
                ];
            }
            else {
                w3cCapabilities.firstMatch[0][defaultCapKey] = defaultCapValue;
            }
        }
    }
    try {
        desiredCaps = (0, base_driver_1.processCapabilities)(w3cCapabilities, constraints, true);
    }
    catch (error) {
        logger_1.log.info(`Could not parse W3C capabilities: ${error.message}`);
        return {
            desiredCaps,
            processedW3CCapabilities,
            error: error,
        };
    }
    processedW3CCapabilities = {
        alwaysMatch: { ...insertAppiumPrefixes(desiredCaps) },
        firstMatch: [{}],
    };
    return {
        desiredCaps,
        processedW3CCapabilities,
    };
}
/**
 * Prefixes capability keys with `appium:` where appropriate.
 */
function insertAppiumPrefixes(caps) {
    return (0, utils_1.mapKeys)(caps, (_, key) => STANDARD_CAPS_LOWERCASE.has(String(key).toLowerCase()) || String(key).includes(':')
        ? String(key)
        : `${W3C_APPIUM_PREFIX}:${key}`);
}
/**
 * Removes `appium:` prefix from capability keys.
 */
function removeAppiumPrefixes(caps) {
    return (0, utils_1.mapKeys)(caps, (_, key) => removeAppiumPrefix(String(key)));
}
/**
 * Pulls Appium settings from capabilities (mutates caps). Supports
 * `settings[key]: value` and `settings: { key: value }`.
 *
 * @returns Parsed settings object; empty if none found.
 */
function pullSettings(caps) {
    if (!support_1.util.isPlainObject(caps) || support_1.util.isEmpty(caps)) {
        return {};
    }
    const result = {};
    const singleSettings = {};
    for (const [key, value] of Object.entries(caps)) {
        let match;
        if (/^(s|appium:s)ettings$/.test(key) && support_1.util.isPlainObject(value)) {
            Object.assign(result, value);
            delete caps[key];
        }
        else if ((match = /^(s|appium:s)ettings\[(\S+)\]$/.exec(key))) {
            singleSettings[match[2]] = value;
            delete caps[key];
        }
    }
    if (!support_1.util.isEmpty(singleSettings)) {
        Object.assign(result, singleSettings);
    }
    return result;
}
function removeAppiumPrefix(key) {
    const prefix = `${W3C_APPIUM_PREFIX}:`;
    return key.startsWith(prefix) ? key.substring(prefix.length) : key;
}
/** True when `caps` has an own property matching `capKey` (with or without `appium:` prefix). */
function hasOwnCapability(caps, capKey) {
    const targetKey = removeAppiumPrefix(capKey);
    return Object.keys(caps).some((key) => removeAppiumPrefix(key) === targetKey);
}
